src/claude_code_sdk/_internal/query.py

```diff
@@ -184,15 +184,35 @@ async def _handle_control_request(self, request: SDKControlRequest) -> None:
                 if not self.can_use_tool:
                     raise Exception("canUseTool callback is not provided")
 
-                response_data = await self.can_use_tool(
+                # Import here to avoid circular dependency
+                from ..types import ToolPermissionContext, ToolPermissionResponse
+
+                context = ToolPermissionContext(
+                    signal=None,  # TODO: Add abort signal support
+                    suggestions=permission_request.get("permission_suggestions", [])
+                )
+
+                response = await self.can_use_tool(
                     permission_request["tool_name"],
                     permission_request["input"],
-                    {
-                        "signal": None,  # TODO: Add abort signal support
-                        "suggestions": permission_request.get("permission_suggestions"),
-                    },
+                    context
                 )
 
+                # Convert ToolPermissionResponse to expected dict format
+                if isinstance(response, ToolPermissionResponse):
+                    response_data = {
+                        "allow": response.allow
+                    }
+                    if response.input is not None:
+                        response_data["input"] = response.input
+                    if response.reason is not None:
+                        response_data["reason"] = response.reason
+                elif isinstance(response, dict):
```

Comment:
Is this elif necessary? We just introduced this, so we should be able to force users to use the cleanest interface?